mqtt/mqtt-bridge/src/persist/storage/ring_buffer/mod.rs

```diff
@@ -0,0 +1,693 @@
+            max_file_size,
+            order: AtomicUsize::default(),
+            pointers: Arc::new(file_pointers),
+            waker: None,
+        }
+    }
+
+    fn insert(&mut self, publication: Publication) -> StorageResult<Key> {
+        let timer = Instant::now();
+        let data = binary_serialize(&publication)?;
+        let data = Data::new(data);
+        let serialized_data = binary_serialize(&data)?;
+
+        let data_size = serialized_data.len();
+        let data_hash = calculate_hash(&data);
+
+        let write_index = self.pointers.write.load(Ordering::SeqCst);
+        let read_index = self.pointers.read.load(Ordering::SeqCst);
+        let order = self.order.load(Ordering::SeqCst);
+        let key = write_index;
+
+        let block_header =
+            BlockHeaderWithHash::new(data_hash, data_size, order, read_index, write_index);
+        let serialized_block_header = binary_serialize(&block_header)?;
+        let block_size = serialized_block_size()?;
+
+        let total_size = block_size + data_size;
+        // Check to see if we might corrupt data if we write, if so, return pending.
+        // As trying later might be successful.
+        if write_index < read_index {
+            if write_index + total_size > read_index {
+                return Err(StorageError::RingBuffer(RingBufferError::Full));
+            }
+        }
+
+        let start = write_index;
+        let end = start + block_size;
+
+        // Check if an existing block header is present. If the block is there
+        // and if the overwrite flag is not set then we shouldn't write.
+        let result = deserialize_block(self.inner.clone(), start, end);
+        if result.is_ok() {
+            let deserialized_block = result.unwrap();
+            let should_not_overwrite = deserialized_block.inner().should_not_overwrite();
+            if should_not_overwrite {
+                return Err(StorageError::RingBuffer(RingBufferError::Full));
+            }
+        }
+
+        let should_flush = should_flush(self.flush_options, &self.flush_state);
+
+        write_data_maybe_wrap_around(
+            self.inner.clone(),
+            start,
+            end,
+            &serialized_block_header,
+            self.max_file_size,
+            should_flush,
+        )?;
+        let start = end;
+        let end = start + data_size;
+        write_data_maybe_wrap_around(
+            self.inner.clone(),
+            start,
+            end,
+            &serialized_data,
+            self.max_file_size,
+            should_flush,
+        )?;
+
+        self.pointers
+            .write
+            .store(end % self.max_file_size, Ordering::SeqCst);
+
+        self.order.store(order + 1, Ordering::SeqCst);
+
+        self.flush_state
+            .update(1, total_size, timer.elapsed().as_millis() as usize);
+        if should_flush {
+            self.flush_state.reset(&self.flush_options);
+        }
+
+        self.wake_up_task();
```

Comment:
not a big deal, but we might call the waker  right after we whore the `write` pointer and before flush.

Reply:
I feel there is a typo above :) Yeah, can move up

Reply:
0_0 oops